### ISSUE TITLE
listen kills parallel

### ISSUE THREAD

**Issue by akerbos** [guard/listen#169]
I execute compilation tasks in parallel with `Parallel.map` from [`grosser/parallel`](https://github.com/grosser/parallel/), each of which writes files to the directory `listen` watches. These files are ignored, but I guess they trigger events anyway. Anyway, after the parallel `map` is done, the program hangs for a while and sometimes throws an exception:

```
/celluloid-0.15.2/lib/celluloid/mailbox.rb:100:in `lock': deadlock detected (fatal)
...
from /.../celluloid-0.15.2/lib/celluloid.rb:131:in `block in register_shutdown'
from /.../parallel-0.9.1/lib/parallel.rb:291:in `fork'
...
```

(also reported as grosser/parallel#88)

Without `listen`, I can not reproduce the error.

Both projects use celluloid and that's where the error comes from; is this related to #163?

**Comment by akerbos** [comment]
With 2.10 and parallel 1.4.1 I still get errors. These days, it's `ERROR -- : Couldn't cleanly terminate all actors in 10 seconds!`. Can not reproduce without listen.

**Comment by e2** [comment]
@akerbos - this should be fixed in the 0.16.0-stable branch of Celluloid (Celluloid didn't handle fork() gracefully).

The backported patch is here: https://github.com/celluloid/celluloid/pull/523

**Comment by akerbos** [comment]
@e2 Wow, thanks for the lightning reply! Here I am, revisiting this after over one year, and you answer within hours. Awesome!

To be clear, since I'm on 0.16.0 (as per `gem list`): am I doing something wrong, or do I simply need an update of `celluloid` beyond 0.16.0 which is yet to be released?
(My use of `parallel` in ltx2any is now somewhat clean ([here](https://github.com/akerbos/ltx2any/blob/master/lib/Extension.rb), lines 76-85) but the use of `listen` is not ([here](https://github.com/akerbos/ltx2any/blob/master/ltx2any.rb), lines 207-255), so I'm nor sure if you want to/can dig through what I do if the former is the case.)

**Comment by e2** [comment]
@akerbos - put the following in your Gemfile (and use Bundler if you can - I don't know if rubygems supports git sources yet):

```ruby
gem 'celluloid', github: 'celluloid/celluloid', branch: '0-16-stable'
```

Meanwhile, I've requested a release of 0.16.1 with this here: https://github.com/celluloid/celluloid/issues/612

**Comment by akerbos** [comment]
Thanks for the hint! I don't use any of these tools (maybe should), so I'll just wait for the next release.

For what it's worth, I managed to build an [MWE](https://gist.github.com/akerbos/70e294bfb7aea943dbea) this time. Can you confirm the problem and that it goes away with the new version?

**Comment by e2** [comment]
@akerbos - it works for me, but you can test it yourself:

1. `git clone https://github.com/celluloid/celluloid.git`
2. switch branches: `git checkout 0-16-stable origin/0-16-stable`
3. revert the reverted patch: `git revert 04af8e7`
3. build the gem with `bundle exec rake build`
4. install it wherever you need: `gem install pkg/celluloid-0.16.0.gem`.

**Comment by akerbos** [comment]
Maybe it's my Git version (1.9.1) but I had to `pull origin 0-16-stable` and then `checkout 0-16-stable` instead of your step 2. Hope I got the right thing now.  
Also, I needed a `bundle install` for build essentials, apparently. First-timer thing, I guess.

Anyway, I can confirm that the issue seems to be eliminated, both with the MWE and my "production" system.

Thank you, both for the fix and your elaborate responses! You have been very helpful when you did not need to be; I appreciate that.

**Comment by e2** [comment]
@akerbos - sorry that should've been: `git checkout -b 0-16-stable origin/0-16-stable`

I'd suggest putting the above `Gemfile` into your project dir and running your scripts with Bundler, e.g. `bundle exec myscript.rb` - since Bundler will pretty much do all those above steps for you (and the gems will work with your dependency management).

Let me know if something doesn't work (don't hesitate to open a new issue) - I would've closed this earlier if I had noticed the `fork()` statement above. And the MWE was *very* helpful for testing - thanks.

Cool project you have there, btw. I've often just hacked together Makefiles to do that kind of stuff but I was never happy with the results.
